### DISCUSSION TITLE
bevy_ecs in a wasm context with no web crypto APIs

### DISCUSSION THREAD

**Discussion by Zynh0722** [bevyengine/bevy#10042] [Q&A]
Hey! I had an idea to use bevy's ecs as a logic controller for my screeps bot, as I think it might be an interesting abstraction to play with.

That aside, my code is running in a context provided by the node vm module, on node v8.9.3, and so far as I can tell, I don't get access to the web crypto apis.

I'll link my [current repo](https://github.com/Zynh0722/screeps/blob/ecs-test/src/lib.rs) just to show you what I have, but in essence I just create a static world and attempt to run a single system schedule.

And unfortunately I get a panic.
```
Error { internal_code: 2147483655, description: "Web Crypto API is unavailable" }
```

idk if it is even possible due to id generation or some other wizardry behind bevy that I don't truly understand, but if anyone has a clue on a direction in the next step, that would be absolutely fantastic (even if it is just make my own domain specific ecs 😩)

**Comment by alice-i-cecile** [comment]
Interesting, I wonder where we're calling RNG functions internally 🤔 Can you try to pare it down to an even more minimal reproduction?

1. Can you run any Rust code?
2. Can you use `bevy::utils::HashMap`?
3. Can you create a `World`?
4. Can you run an empty schedule?
5. Can you run a single system without a schedule?
6. Can you put that system in the schedule and run it?

**Reply by Zynh0722** [reply]
1. Yes! my pre ecs code runs quite well from the main branch of that same repository
2. I don't actually have bevy itself as a dependency, just bevy_ecs
3. It seems not. 
 - Simply running below in the main loop results in the same panic
```rs
let _world = World::default();
```
4. dependent on 3
5. dependent on 3
6. dependent on 3

**Reply by Zynh0722** [reply]
For completion, I switched to the root bevy dependency (much to the chagrin of rust-analyzer), and I can use `bevy::utils::HashMap`

At least making a default one each tick works 
```rs
let _map: bevy::utils::HashMap<String, String> = bevy::utils::HashMap::default();
````

E: switch to full codeblock

**Comment by Bluefinger** [comment]
Node 8.9.3 is pretty old, and won't have access to the web crypto API that is being used by `getrandom`'s js bindgen. You'll require at least Node 15 IIRC. Any part of bevy's code that relies on RNG seeded from OS sources via `getrandom` will make use of the Web Crypto API. This will include the most recent versions of `fastrand` with the `js` feature enabled.

**Reply by Zynh0722** [reply]
this doesn't totally surprise me, 2017 is indeed ancient in node terms, going to see if there is any possible workaround lmao

unfortunately the runtime my code runs in is outside of my control

**Reply by Bluefinger** [reply]
Fair, but keep in mind that for the std `HashMap`/`DefaultHasher` or basically anything using `RandomState` to initialise will **not** actually be random on unsupported platforms (plain WASM is not supported for getting OS/system level RNG sources, WASI is required for this purpose). It's a big gotcha if you are planning to use these on WASM without `getrandom` and a reasonably recent runtime that gives a good entropy source.

Your workaround might require constructing your own entropy sourcing, though `getrandom` does offer the means to do so for platforms it doesn't support directly.

**Comment by lee-orr** [comment]
One possible approach is creating a JS shim emulating the web crypto API. Because all the interaction is happening via javascript, you might be able to make bevy think that API exists if you add a fully-JS version into the global scope.

**Reply by lee-orr** [reply]
Something like https://www.npmjs.com/package/webcrypto-shim
